core/src/main/java/org/apache/iceberg/ManifestFilterManager.java

```diff
@@ -81,6 +81,7 @@ public String partition() {
 
   // cache filtered manifests to avoid extra work when commits fail.
   private final Map<ManifestFile, ManifestFile> filteredManifests = Maps.newConcurrentMap();
+  private final Set<String> manifestsWithDeletedFiles = Sets.newConcurrentHashSet();
```

Comment:
I actually thought it would be called like `deleteFileManifests` and behave similar to `deleteFilePartitions`. We will keep a flag similar to `hasPathOnlyDeletes` to know if `deleteFileManifests` can be trusted. The main idea is by knowing that all files to be deleted had `manifestLocation` set correctly, we can rely on the collected manifest set to reliably tell if a particular manifest has to be rewritten or not. In the current version of the PR, we only skip opening manifests that we have matches for. However, we still read all other manifests that don't have matches. If we know that  `deleteFileManifests` can be trusted, we can completely avoid reading manifests to determine what to rewrite.

Does it make sense?

Reply:
Thanks for the explanation! 

>However, we still read all other manifests that don't have matches. If we know that deleteFileManifests can be trusted, we can completely avoid reading manifests to determine what to rewrite.

Hm I'll double check if I'm missing something but I think this should already be happening in this PR. Before opening any manifest we check canContainDeletedFiles(ManifestFile manifest). If this is false, we skip the opening up of the manifest.

Looking at the internal logic of this method:

https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/ManifestFilterManager.java#L329

Let me know if there's a flaw in my reasoning:

1. The case where `deleteFileManifests` can be trusted is when there's no row expression for deletion, when `hasPathOnlyDeletes` is false, when there are no dropped partitions and when there are deletes that have aged out.

2. If the file to be deleted is referenced in a manifest the current set will contain that and we return true for that manifest.

3. Given the statement in 1 where there's no row expression, no path only deletes, no dropped partitions, aged out deletes and there's no referenced manifest we would end up returning false here https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/ManifestFilterManager.java#L362.

4. Then we would skip going through the manifest and add it to the filtered set/return https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/ManifestFilterManager.java#L304

Reply:
I took another look at `canContainDeletedFiles`.

```
private boolean canContainDeletedFiles(ManifestFile manifest) {
  if (manifestsWithDeletedFiles.contains(manifest.path())) {
    return true;
  }

  ...
}
```

I think the branch only covers cases when we know it is there. If the manifest location is not in the set, we proceed further with reading the manifest.

Reply:
I discussed offline and stepped through `canContainDeletedFiles` with @aokolnychyi, one branch in `canContainDeletedFiles` that leads to false positives in the current PR is https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/ManifestFilterManager.java#L352. 

Let's say the majority of manifests don't need to be rewritten and the table is unpartitioned we'd return true [here](https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/util/ManifestFileUtil.java#L109). 

The same principle applies if the table is partitioned, it still depends on how the deletes are grouped and how the manifests are grouped.

To solve against false positives, locally I added a flag which explicitly tracks if all the delets being performed are on files which have referenced manifests. If that is the case, we know we can use the set as a source of truth; if the referenced location is not in the set then we know with certainty the manifest doesn't have to be rewritten.

Reply:
Opening back up this thread in case we wanted to discuss the field name since I haven't changed that yet:`manifestsWithDeletedFiles` vs `deleteManifests`.

This change is keeping track of both data and delete manifests which have entries which are deleted (hence the name `manifestsWithDeletedFiles`). When we go to prune out which manifests *cannot* contain any deleted entries, that's when an additional check on delete manifests is performed since cherrypick overwrite cases make it so that detecting data manifests cannot contain deleted entries is not possible. 

I could also only keep track of specifically the delete manifests and then rename to `deleteManifests` but the advantage that gets lost is the ability to detect if a data manifest has an entry which is deleted, we has to go through the old path.